ArrayOrder.md

```diff
@@ -0,0 +1,202 @@
+where `2*3 = 6`. There is no grouping to be done for a dimensions of stride 1,
+so the first step is to join elements into groups of 2 (the
+second stride):
+
+`[(0, 1), (2, 3), (4, 5), (6, 7), (8, 9), (10, 11), (12, 13), (14, 15), (16, 17), (18, 19), (20, 21), (22, 23)]`
+
+Next, group elements of the new list (which are themselves groups) into
+groups of 3 (the largest stride).
+
+`[((0, 1), (2, 3), (4, 5)), ((6, 7), (8, 9), (10, 11)), ((12, 13), (14, 15), (16, 17)), ((18, 19), (20, 21), (22, 23))]`
+
+Notice:
+* The element adjacent to `0` in the inner group is `1`, hence stride `1`.
+* The element adjacent to `0` in the intermediate grouping is `2`, hence stride `2`.
+* The element adjacent to `0` in the outer grouping is `6`, hence stride `6`.
+
+</details>
+
+
+### C- and F-order
+
+* **Define:** C-order indexes multi-dimensional arrays such that the **last** index has stride 1.
+* **Define:** F-order indexes multi-dimensional arrays such that the **first** index has stride 1.
+
+These terms come from conventions for storing arrays in the C and Fortran programming languages.
+
+### array size
+
+The size of a multidimensional array is described by a list of sizes per
+dimension. For example: `[3, 5, 7].` In this example, the *first*
+dimension has size `3`, the *last* dimension has size `7`.
+
+If this array is indexed using C-order, then the last index has stride
+`1`. As a result, the middle index will have stride `7`, and the *first*
+dimension will have stride `7*5 = 35`. 
+
+Consider again an array of size `[3, 5, 7]`, but using F-order
+indexing. Again, the *first* dimension has size `3`, the *last*
+dimension has size `7`. Now, however, using F-order, the 
+*first* dimension will have stride `1`, the *second* dimension will have
+stride `3`, and the *third* dimension will have stride `3*5=15`.
+
+## Interpretation of multi-dimensional arrays
+
+The following sections require adding an interpretation or semantics to
+the dimensions of a multi-dimensional array. We'll discuss interpreting
+arrays as matrices and images.
+
+### row- and column-major
+
+Matrices are usually represented as a 2D array of numbers. Horizontal
+groupings of these numbers are called "rows" and vertical groupings are
+called "columns." In mathematics, the entries of a matrix $A$ are
+denoted $a_{ij}$. Rows of the matrix are indexed by $i$, the
+"first" or "left" index, and columns of the matrix are indexed by $j$,
+the "second" or "right" index.
+
+* **Universal Matrix Convention**: Left indexes refer to rows,
+  right indexes refer to columns.
+
+The terms row- and column-major derive for the storage of matrices.
+Defining these terms first depends on first agreeing which index (left
+or right) refers to rows vs columns for matrices in mathematics.
+
+* **Define:** Arrays storing matrices in "row-major" give columns stride 1. 
+* **Define:** Arrays storing matrices in "column-major" give rows stride 1. 
```

Comment:
Is the use of rows and columns correct here? As far as I understand the concepts introduced here, "row-major" stores rows contiguous in memory, i.e., gives rows stride 1. I might be missing something, though.

Reply:
Yea, this is backwards. We also may need to put a unit on `1`.

In Julia, column-major, the strides are as follows, given in terms of the type (`Float64` aka `double`).
```julia
julia> strides(zeros(5,6,7))
(1, 5, 30)
```

In Python, the strides are given in terms of bytes.
```python
In [5]: np.zeros((5,6,7)).strides
Out[5]: (336, 56, 8)
```

Reply:
"row is contiguous in memory" is equivalent to "column index has stride one"

Reply:
I see what you mean by this comment. I interpreted "gives columns stride 1" as "iterating within a column (i.e., changing the row index) has stride 1". In my view, my interpretation is consistent with your definition above:
>  the stride of a dimension is the (positive) step in the flat array that corresponds to the adjacent element along that dimension.

What about replacing "columns" and "rows" by "second" and "first" index, respectively?